ICE: thread 'rustc' panicked at 'normalization works on validated constants...'

I tried updating [Tock](github.com/tock/tock) to today's nightly and got the following clippy internal error:

```
    Checking nrf52_components v0.1.0 (/home/runner/work/tock/tock/boards/nordic/nrf52_components)
thread 'rustc' panicked at 'normalization works on validated constants: InterpErrorInfo(InterpErrorInfoInner { kind: using uninitialized data, but this operation requires initialized memory, backtrace: None })', compiler/rustc_const_eval/src/const_eval/eval_queries.rs:138:36
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.65 (c2804e6 2022-09-07)

query stack during panic:
#0 [try_destructure_mir_constant] destructuring mir constant
#1 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `litex_sim`
warning: build failed, waiting for other jobs to finish...
make: *** [Makefile:347: ci-job-clippy] Error 101
Error: Process completed with exit code 2.
```

The full CI output is here: https://github.com/tock/tock/runs/8251506744?check_suite_focus=true

The pull request to update the nightly is here: https://github.com/tock/tock/pull/3168

The crate it appears to be failing on is here: https://github.com/tock/tock/tree/master/boards/nordic/nrf52_components. Note, the PR does not change that particular crate.

--- a/clippy_lints/src/non_copy_const.rs
+++ b/clippy_lints/src/non_copy_const.rs
@@ -149,6 +149,9 @@ fn is_value_unfrozen_raw<'tcx>(
             // the fact that we have to dig into every structs to search enums
             // leads us to the point checking `UnsafeCell` directly is the only option.
             ty::Adt(ty_def, ..) if ty_def.is_unsafe_cell() => true,
+            // As of 2022-09-08 miri doesn't track which union field is active so there's no safe way to check the
+            // contained value.
+            ty::Adt(def, ..) if def.is_union() => false,
             ty::Array(..) | ty::Adt(..) | ty::Tuple(..) => {
                 let val = cx.tcx.destructure_mir_constant(cx.param_env, val);
                 val.fields.iter().any(|field| inner(cx, *field))

--- a/tests/ui/crashes/ice-9445.rs
+++ b/tests/ui/crashes/ice-9445.rs
@@ -1,0 +1,3 @@
+const UNINIT: core::mem::MaybeUninit<core::cell::Cell<&'static ()>> = core::mem::MaybeUninit::uninit();
+
+fn main() {}

